### PR TITLE
refactor: results map and match creation

### PR DESCRIPTION
**Pull request by Disservin**
wip, needs some careful eyes if everything still makes sense, i.e. stats collection, elo calculation, penta stats, correct flip of engines etc

### CODE REVIEW

**Comment by gahtan-syarif** [comment]
let me know when the rewrite is done so i can do a side by side test with my branch

**Comment by Disservin** [comment]
@gahtan-syarif give it a go

**Comment by gahtan-syarif** [comment]
> @gahtan-syarif give it a go

alright hold up

**Comment by gahtan-syarif** [comment]
annotation doesnt match result: `1/2-1/2 {Black wins by adjudication}`

**Comment by gahtan-syarif** [comment]
final result this branch:
```
Finished game 46 (Base-ee6fc7e38b4aeef44862159215a56d97122f59a0 vs New-ee6fc7e38b4aeef44862159215a56d97122f59a0): 1-0 {White wins by adjudication}
Score of New-ee6fc7e38b4aeef44862159215a56d97122f59a0 vs Base-ee6fc7e38b4aeef44862159215a56d97122f59a0: 10 - 10 - 26 [0.500] 46
Elo difference: -0.00 +/- 67.02, LOS: 50.00 %, DrawRatio: 56.52 %
```

cutechess and my branch:
```
Score of New-ee6fc7e38b4aeef44862159215a56d97122f59a0 vs Base-ee6fc7e38b4aeef44862159215a56d97122f59a0: 19 - 19 - 8  [0.500] 46
Elo difference: -0.00 +/- 93.44, LOS: 50.00 %, DrawRatio: 17.39 %
Finished match
```

**Comment by gahtan-syarif** [comment]
[console.txt](https://github.com/user-attachments/files/16202697/console.txt)

**Comment by vondele** [comment]
I also gave this branch a try via the fishtest worker, and we still have the following which causes an assert:
```
Score of New-8d1e41458e vs Base-68e1e9b381: 109 - 248 - 346 [0.401] 703
Finished game 695 (New-8d1e41458e vs Base-68e1e9b381): 1-0 {White wins by adjudication}
Finished game 718 (Base-68e1e9b381 vs New-8d1e41458e): 1-0 {White wins by adjudication}
Score of New-8d1e41458e vs Base-68e1e9b381: 110 - 248 - 346 [0.402] 704
```
(i.e. two finished games and score only increments by 1), but not sure if this branch was supposed to fix it

File: app/src/matchmaking/stats.hpp [hunk 1]
```diff
@@ -18,9 +18,9 @@ class Stats {
         penta_WW = ww;
     }
     Stats(const MatchData &match_data) {
-        if (match_data.players.first.result == chess::GameResult::WIN) {
+        if (match_data.players.white.result == chess::GameResult::WIN) {
             wins++;
-        } else if (match_data.players.first.result == chess::GameResult::LOSE) {
+        } else if (match_data.players.black.result == chess::GameResult::LOSE) {
```
Comment (gahtan-syarif) [inline app/src/matchmaking/stats.hpp:23]: this should be white i think?

**Comment by Disservin** [comment]
> I also gave this branch a try via the fishtest worker, and we still have the following which causes an assert:
> 
> ```
> Score of New-8d1e41458e vs Base-68e1e9b381: 109 - 248 - 346 [0.401] 703
> Finished game 695 (New-8d1e41458e vs Base-68e1e9b381): 1-0 {White wins by adjudication}
> Finished game 718 (Base-68e1e9b381 vs New-8d1e41458e): 1-0 {White wins by adjudication}
> Score of New-8d1e41458e vs Base-68e1e9b381: 110 - 248 - 346 [0.402] 704
> ```
> 
> (i.e. two finished games and score only increments by 1), but not sure if this branch was supposed to fix it

No ;) this is only a hopefully NFC refactor aiming to make all logic related to
> In cutechess, odd games are always played in white pov while even games are played in black pov.

more robust (black pov is wrong thing to say all output is from white pov but just engines switch place)

File: app/src/matchmaking/stats.hpp (hunk 1 above)
Comment (Disservin) [reply app/src/matchmaking/stats.hpp:23]: Ah yeah

**Comment by Disservin** [comment]
@gahtan-syarif can you check again?

**Comment by vondele** [comment]
Not yet OK, relative to cutechess:
```
< Finished game 21 (sf_1 vs sf_2): 1-0 {White wins by adjudication}
---
> Finished game 21 (sf_1 vs sf_2): 0-1 {Black wins by adjudication}
```
In some cases the difference might be irrelevant:
```
< Finished game 72 (sf_2 vs sf_1): 1/2-1/2 {Draw by adjudication}
---
> Finished game 72 (sf_2 vs sf_1): 1/2-1/2 {Draw by 3-fold repetition}
```
Or
```
< Finished game 87 (sf_1 vs sf_2): 0-1 {Black mates}
---
> Finished game 87 (sf_1 vs sf_2): 0-1 {White got checkmated}
```

Obtained with:
```bash
binary=./cutechess-cli
binary=./fast-chess

for binary in ./cutechess-cli ./fast-chess
do

echo " ============= $binary ============== "

(

$binary         -recover -repeat -games 2 -rounds 100 -tournament gauntlet\
                -pgnout $binary-out.pgn\
                -srand $RANDOM  -resign movecount=3 score=600 -draw movenumber=34 movecount=8 score=20\
                -variant standard -concurrency 4 -openings file=book.epd format=epd order=sequential \
                -engine name=sf_1 tc=100+0.1 depth=10 cmd=./sf_1 dir=.\
                -engine name=sf_2 tc=100+0.1 depth=12 cmd=./sf_2 dir=.\
                -each proto=uci option.Threads=1

) | grep "Finished game" | sort -n -k 3 > $binary-out.finished.

done
```
and
```bash
head -n 100 UHO_Lichess_4852_v1.epd | awk '{print $1,$2,$3,$4}' > book.epd
```

**Comment by vondele** [comment]
OK, the issue for the different result is actually something else, looks like fast-chess reads the book differently, leading to a different game. I'll make a different bug report for that.

**Comment by vondele** [comment]
as the issue https://github.com/Disservin/fast-chess/issues/559 was the only issue with a different result, this patch might be right. The other differences are more different wording, and I assume are in part taken care of by https://github.com/Disservin/fast-chess/pull/552

**Comment by gahtan-syarif** [comment]
> @gahtan-syarif can you check again?

output the same as cutechess:
```
Finished game 46 (Base-ee6fc7e38b4aeef44862159215a56d97122f59a0 vs New-ee6fc7e38b4aeef44862159215a56d97122f59a0): 1-0 {White wins by adjudication}
Score of New-ee6fc7e38b4aeef44862159215a56d97122f59a0 vs Base-ee6fc7e38b4aeef44862159215a56d97122f59a0: 19 - 19 - 8 [0.500] 46
Elo difference: -0.00 +/- 93.44, LOS: 50.00 %, DrawRatio: 17.39 %
Saved results.
Finished match
```
ill check the penta output now

**Comment by gahtan-syarif** [comment]
@Disservin when played at tc odds, the displayed tc seems to switch around:
```
--------------------------------------------------
Results of New-ee6fc7e38b4aeef44862159215a56d97122f59a0 vs Base-ee6fc7e38b4aeef44862159215a56d97122f59a0 (7000 nodes - 5000 nodes, 1t, 16MB, UHO_Lichess_4852_v1.epd):
Elo: inf +/- nan, nElo: inf +/- nan
LOS: 100.00 %, DrawRatio: 0.00 %, PairsRatio: inf
Games: 6, Wins: 6, Losses: 0, Draws: 0, Points: 6.0 (100.00 %)
Ptnml(0-2): [0, 0, 0, 0, 3], WL/DD Ratio: -nan
--------------------------------------------------
Started game 10 of 46 (Base-ee6fc7e38b4aeef44862159215a56d97122f59a0 vs New-ee6fc7e38b4aeef44862159215a56d97122f59a0)
Finished game 7 (New-ee6fc7e38b4aeef44862159215a56d97122f59a0 vs Base-ee6fc7e38b4aeef44862159215a56d97122f59a0): 1-0 {White wins by adjudication}
Started game 11 of 46 (New-ee6fc7e38b4aeef44862159215a56d97122f59a0 vs Base-ee6fc7e38b4aeef44862159215a56d97122f59a0)
Finished game 9 (New-ee6fc7e38b4aeef44862159215a56d97122f59a0 vs Base-ee6fc7e38b4aeef44862159215a56d97122f59a0): 1-0 {White wins by adjudication}
Started game 12 of 46 (Base-ee6fc7e38b4aeef44862159215a56d97122f59a0 vs New-ee6fc7e38b4aeef44862159215a56d97122f59a0)
Finished game 8 (Base-ee6fc7e38b4aeef44862159215a56d97122f59a0 vs New-ee6fc7e38b4aeef44862159215a56d97122f59a0): 0-1 {Black wins by adjudication}
--------------------------------------------------
Results of New-ee6fc7e38b4aeef44862159215a56d97122f59a0 vs Base-ee6fc7e38b4aeef44862159215a56d97122f59a0 (5000 nodes - 7000 nodes, 1t, 16MB, UHO_Lichess_4852_v1.epd):
Elo: inf +/- nan, nElo: inf +/- nan
LOS: 100.00 %, DrawRatio: 0.00 %, PairsRatio: inf
Games: 8, Wins: 8, Losses: 0, Draws: 0, Points: 8.0 (100.00 %)
Ptnml(0-2): [0, 0, 0, 0, 4], WL/DD Ratio: -nan
--------------------------------------------------
```
in this case `New` should always be 7000 nodes and `Base` 5000 nodes

**Comment by gahtan-syarif** [comment]
@Disservin 
```
fast-chess.exe -config file=config.json
Loading config file: config.json
Error while reading option "-config" with value "file=config.json"
Reason: [json.exception.out_of_range.403] key 'key' not found
```

**Comment by Disservin** [comment]
both things fixed, vondeles are for your pr/later pr

**Comment by gahtan-syarif** [comment]
alright ill check

**Comment by gahtan-syarif** [comment]
everything looks good @Disservin if i missed anything it would be pretty minor

**Pull request merged**